File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return dfs(root,INT_MIN,INT_MAX);
    }

    bool dfs(TreeNode* node,int low,int high){
        if(!node) return true; 

        if(node->val<=low || node->val>=high) return false; 

        bool lst = dfs(node->left,low,node->val); 
        if(!lst) return false;
        bool rst = dfs(node->right,node->val,high); 
        return lst&&rst;

    }
};

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self,
        root: Optional[TreeNode],mini=None,
        maxi=None) -> bool:
        
        if root is None: 
            return True  
        if mini and maxi and (root.val<=mini.val or root.val>=maxi.val): 
            return False
        
        return self.isValidBST(root.left,mini,root) and self.isValidBST(root.right,root,maxi)
